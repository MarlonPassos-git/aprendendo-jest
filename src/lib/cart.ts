import find  from 'lodash/fp/find'

type IProduct = {
  title: string
  price: number
}

type IItem = {
	product: IProduct
	quantity: number
}

/**
 * Componente responsavel por controlar todo o fuxo do carrinho de compras
 */
export class Cart {
	private totalItems: number = 0
	private listProducts: IItem[] = []
	private totalPrice: number = 0
	private totalProduct: number = 0

	constructor() {}

	public getTotalItens(): number {
		return this.totalItems
	}

	public add(product: IItem): void {
		++this.totalItems
		this.listProducts.push(product)
		this.updateTotalPrice()
		this.updateTotalProducts()
	}

	public getListProducts(): IItem[] {
		return this.listProducts
	}

	public getTotalPrice(): number {
		return this.totalPrice
	}

	public getTotalProducts(): number {
		return this.totalProduct
	}

	private updateTotalPrice(): void {
		const totalPrice = this.listProducts.reduce(
			(tot, { product: { price }, quantity }) => {
				return price * quantity + tot
			},
			0,
		)

		this.totalPrice = totalPrice
	}

	private updateTotalProducts(): void { 
		const totalProducts =  this.listProducts.reduce((acc, { quantity }) => {
			return acc + quantity
		}, 0)

		this.totalProduct = totalProducts
	}
}
