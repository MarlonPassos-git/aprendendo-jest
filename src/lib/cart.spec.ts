import { Cart } from './cart'

describe('cart1', () => {
  let cart: Cart
  const item1 = {
		product: {
			title: 'Product 1',
			price: 1000,
		},
		quantity: 2,
	}

  const item2 = {
		product: {
			title: 'Product 2',
			price: 12345,
		},
		quantity: 3,
	}

  const item3 = {
		product: {
			title: 'Product ',
			price: 5000,
		},
		quantity: 10,
	}
  beforeEach(() => {
    cart = new Cart()
  })
  
  it('should return 0 when getTotal() is executed in a newly created ', () => {
    expect(cart.getTotalItens()).toEqual(0)
  })

  it('should empty list when started', () => {
    expect(cart.getListProducts()).toEqual([])
  })

  it('should be able to add an item to cart', () => {
		cart.add(item1)

    expect(cart.getListProducts()).toEqual([item1])
    expect(cart.getTotalItens()).toEqual(1)
    expect(cart.getTotalPrice()).toEqual(2000)
    expect(cart.getTotalProducts()).toEqual(2)

  })
  
  it('should be able to add two products at the same time', () => {
    cart.add(item1)
    cart.add(item2)

    expect(cart.getListProducts()).toEqual([item1, item2])
    expect(cart.getTotalItens()).toEqual(2)
    expect(cart.getTotalPrice()).toEqual(12345 * 3 + 1000 * 2)
    expect(cart.getTotalProducts()).toEqual(5)

  })

  it('if the item exists in the cart and is re-added, the quantity of the last addition must remain', () => {
    const item1Plus = {
			product: {
				title: 'Product 1',
				price: 1000,
			},
			quantity: 5,
		}
    
    cart.add(item1)
    cart.add(item1Plus)
    
    expect(cart.getListProducts()).toEqual([item1Plus])

  })

  it('should keeo the values consistent when removing a product', () => {
    cart.add(item1)
    cart.add(item2)
    cart.add(item3)
    cart.remove(item1)

    expect(cart.getListProducts()).toEqual([item2])
    expect(cart.getTotalItens()).toEqual(2)
    expect(cart.getTotalPrice()).toEqual(12345 * 3 + 5000 * 10)
    expect(cart.getTotalProducts()).toEqual(13)
  })


})
// ssas 