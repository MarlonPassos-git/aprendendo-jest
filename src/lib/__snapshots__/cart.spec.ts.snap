// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cart1 It should show an empty cart when i add the item and after that it's skein 1`] = `Array []`;

exports[`cart1 It should show an1 empty cart when i add the item and after that it's skein 1`] = `Array []`;

exports[`cart1 Must show a summary of the entire cart 1`] = `
Object {
  "items": 2,
  "price": 55000,
  "produts": Object {
    "list": Array [
      Object {
        "product": Object {
          "price": 1000,
          "title": "Product 1",
        },
        "quantity": 5,
      },
      Object {
        "product": Object {
          "price": 5000,
          "title": "Product ",
        },
        "quantity": 10,
      },
    ],
    "total": 15,
  },
}
`;
